File: 4D4Progr4m5/MergeSortWithTimeComplexity.c
// clock_t is a typedef for long that represents processor time values (clock ticks), typically returned by functions like clock() and converted to seconds using CLOCKS_PER_SEC.
#include<stdio.h>
#include<time.h>
#include<math.h>
#include<stdlib.h>

void mergesort(int *, int, int);
void merge(int *, int ,int, int);

void main()
{
    int n,i;
    int a[20];
    float t;
    printf("Enter the number of elements : ");
    scanf("%d",&n);
    printf("Enter %d elements into the array : ", n);
    for(i=0;i<n;i++)
    {
        scanf("%d",&a[i]);
    }
    clock_t start = clock();
    mergesort(a,0,n-1);
    clock_t end = clock();
    printf("Sorted array : \n");
    for(i=0;i<n;i++)
    {
        printf("%d ",a[i]);
    }
    t=n*(log(n));
    printf("\nTime Complexity %f : ",t);
    return ;

}

void mergesort(int a[], int lb, int ub)
{
    int mid;
    if(lb<ub)
    {
        mid=(lb+ub)/2;
        mergesort(a,lb,mid);
        mergesort(a,mid+1,ub);
        merge(a,lb,mid,ub);
    }
}
void merge(int a[], int lb, int mid, int ub)
{
    int i,j,k;
    int c[20];
    i=lb;
    k=lb;
    j=mid+1;
    while((i<=mid)&&(j<=ub))
    {
        if(a[i]<a[j])
        {
            c[k]=a[i];
            k++;
            i++;
        }
        else
        {
            c[k]=a[j];
            k++;
            j++;
        }
    }
    while(i<=mid)
    {
        c[k]=a[i];
        k++;
        i++;
    }
    while(j<=ub)
    {
        c[k]=a[j];
        k++;
        j++;
    }
    for(i=lb;i<k-1;i++)
    {
        a[i]=c[i];
    }

}

File: 4D4Progr4m5/MergeSortWithTimeComplexity.py
def mergeSort(a, lb, ub):
    #mid = (ub+lb)//2
    if(lb<ub):
        mid = (ub+lb)//2
        mergeSort(a,lb,mid)
        mergeSort(a,mid+1,ub)
        merge(a,lb,mid,ub)

def merge(a, lb, mid, ub):
    i=lb
    k=lb
    j=mid+1
    c=[]
    while i<=mid and j<=ub:
        if a[i]<a[j]:
            c[k]=a[i]
            k+=1
            i+=1
        else:
            c[k]=a[j]
            k+=1
            i+=1
    while i<=mid:
        c[k]=a[i]
        k+=1
        i+=1
    while j<=ub:
        c[k]=a[j]
        k+=1
        i+=1
    for i in range(lb,k-1):
        a[i]=c[i]

import math
n=int(input("enter the number of elements : "))
arr = input("Enter the elements into the array seperated by space : ")
a=[int(i) for i in arr.split()]
sortedArray=mergeSort(a,0,n-1)
print("Sorted Array : ", sortedArray)
t=n*(math.log(n))
print("Time complexity : ",t)


